File: android/src/main/kotlin/com/wultra/android/powerauth/flutter/internal/services/PowerAuthUtilsService.kt
/*
 * Copyright 2025 Wultra s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.wultra.android.powerauth.flutter.internal.services

import com.wultra.android.powerauth.flutter.Errors
import com.wultra.android.powerauth.flutter.WrapperException
import com.wultra.android.powerauth.flutter.internal.core.BasePowerAuthService
import com.wultra.android.powerauth.flutter.internal.core.PowerAuthFlutterService.MethodHandler
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import io.getlime.security.powerauth.core.ActivationCode
import io.getlime.security.powerauth.core.ActivationCodeUtil

internal class PowerAuthUtilsService : BasePowerAuthService(null) {

    override val name = "util"

    companion object ArgKeys {
        const val ACTIVATION_CODE = "activationCode"
        const val ACTIVATION_SIGNATURE = "activationSignature"
        const val CHARACTER = "character"
    }

    private object HandlerNames {
        const val UTIL_PARSE_ACTIVATION_CODE = "parseActivationCode"
        const val UTIL_VALIDATE_ACTIVATION_CODE = "validateActivationCode"
        const val UTIL_VALIDATE_TYPED_CHARACTER = "validateTypedCharacter"
        const val UTIL_CORRECT_TYPED_CHARACTER = "correctTypedCharacter"
    }

    override val handlers by lazy {
        mapOf(
            HandlerNames.UTIL_PARSE_ACTIVATION_CODE to MethodHandler { call, result ->
                parseActivationCode(
                    call,
                    result
                )
            },
            HandlerNames.UTIL_VALIDATE_ACTIVATION_CODE to MethodHandler { call, result ->
                validateActivationCode(
                    call,
                    result
                )
            },
            HandlerNames.UTIL_VALIDATE_TYPED_CHARACTER to MethodHandler { call, result ->
                validateTypedCharacter(
                    call,
                    result
                )
            },
            HandlerNames.UTIL_CORRECT_TYPED_CHARACTER to MethodHandler { call, result ->
                correctTypedCharacter(
                    call,
                    result
                )
            },
        )
    }

    private fun parseActivationCode(call: MethodCall, result: Result) {
        try {
            val activationCodeString: String = call.getRequiredArgument(ACTIVATION_CODE)
            val ac: ActivationCode? = ActivationCodeUtil.parseFromActivationCode(activationCodeString)

            if (ac != null) {
                val response = mutableMapOf<String, String?>(
                    ACTIVATION_CODE to ac.activationCode
                )

                ac.activationSignature?.let { response[ACTIVATION_SIGNATURE] = it }

                result.success(response)
            } else {
                result.error(Errors.EC_INVALID_ACTIVATION_CODE, "Invalid activation code provided.", null)
            }
        } catch (t: Throwable) {
            Errors.error(result, t)
        }
    }

    private fun validateActivationCode(call: MethodCall, result: Result) {
        try {
            val activationCodeString: String = call.getRequiredArgument(ACTIVATION_CODE)

            result.success(ActivationCodeUtil.validateActivationCode(activationCodeString))
        } catch (t: Throwable) {
            Errors.error(result, t)
        }
    }

    private fun validateTypedCharacter(call: MethodCall, result: Result) {
        try {
            val characterInt: Int = call.getRequiredArgument(CHARACTER)
            result.success(ActivationCodeUtil.validateTypedCharacter(characterInt))
        } catch (e: NumberFormatException) {
            Errors.error(result, WrapperException(Errors.EC_WRONG_PARAMETER, "Invalid character format, expected Int", e))
        } catch (t: Throwable) {
            Errors.error(result, t)
        }
    }

    private fun correctTypedCharacter(call: MethodCall, result: Result) {
        try {
            val characterInt: Int = call.getRequiredArgument(CHARACTER)
            val corrected: Int = ActivationCodeUtil.validateAndCorrectTypedCharacter(characterInt)

            if (corrected == 0) {
                result.error(Errors.EC_INVALID_CHARACTER, "Invalid character that cannot be corrected.", null)
            } else {
                result.success(corrected)
            }
        } catch (e: NumberFormatException) {
            Errors.error(result, WrapperException(Errors.EC_WRONG_PARAMETER, "Invalid character format, expected Int", e))
        } catch (t: Throwable) {
            Errors.error(result, t)
        }
    }
}

File: ios/flutter_powerauth_mobile_sdk_plugin/Sources/flutter_powerauth_mobile_sdk_plugin/internal/services/PowerAuthUtilsService.swift
/*
 * Copyright 2025 Wultra s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Flutter
import UIKit
import PowerAuth2
import PowerAuthCore

internal class PowerAuthUtilsService: PowerAuthFlutterService {
    
    typealias Service = PowerAuthUtilsService
    
    // MARK: - PowerAuthFlutterService members
    
    var name: String { "PowerAuthUtils" }
    
    let handlers = [
        "util_parseActivationCode": parseActivationCode,
        "util_validateActivationCode": validateActivationCode,
        "util_validateTypedCharacter": validateTypedCharacter,
        "util_correctTypedCharacter": correctTypedCharacter,
        "util_getEnvironmentInfo": getEnvironmentInfo,
        "util_migrateSharingConfiguration": migrateSharingConfiguration
    ]
    
    fileprivate enum Args: String {
        case character
        case activationCode
        case fromAppGroup
        case toAppGroup
    }
    
    private func validateActivationCode(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        let code: String = try call.requireParameter(Args.activationCode)
        result(PowerAuthActivationCodeUtil.validateActivationCode(code))
    }
    
    private func parseActivationCode(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        let code: String = try call.requireParameter(Args.activationCode)
        if let parsed = PowerAuthActivationCodeUtil.parse(fromActivationCode: code) {
            result([
                "activationCode": parsed.activationCode,
                "activationSignature": parsed.activationSignature
            ])
        } else {
            throw PluginException(.invalidActivationCode, message: "Invalid activation code.")
        }
    }
    
    private func validateTypedCharacter(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        let char: Int = try call.requireParameter(Args.character)
        result(PowerAuthActivationCodeUtil.validateTypedCharacter(UInt32(char)))
    }
    
    private func correctTypedCharacter(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        let char: Int = try call.requireParameter(Args.character)
        let validated = PowerAuthActivationCodeUtil.validateAndCorrectTypedCharacter(UInt32(char))
        result(validated)
    }

    private func getEnvironmentInfo(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        let currentDevice = UIDevice.current
        let mainBundle = Bundle.main
        let mainDictionary = mainBundle.infoDictionary
        let appVersion = mainDictionary?["CFBundleShortVersionString"] as? String
        let appId = mainDictionary?["CFBundleIdentifier"] as? String
        
        result([
            "systemName": currentDevice.systemName,
            "systemVersion": currentDevice.systemVersion,
            
            "applicationVersion": appVersion,
            "applicationIdentifier": appId,
            
            "deviceManufacturer": "apple",
            "deviceId": currentDevice.model
        ])
    }

    /// Migrates the keychain initialization flag (`PowerAuthKeychain_Initialized`) between two
    /// `UserDefaults` suites identified by their app group names.
    ///
    /// This mirrors the migration described in the PowerAuth SDK for iOS Extensions documentation,
    /// but resolves the `UserDefaults` suite directly from the provided app group instead of the
    /// keychain configuration's `keychainAttribute_UserDefaultsSuiteName`. A `nil` app group
    /// resolves to the standard (non-shared) `UserDefaults`.
    ///
    /// See: https://developers.wultra.com/components/powerauth-mobile-sdk/1.9.x/documentation/PowerAuth-SDK-for-iOS-Extensions#userdefaults-migration
    private func migrateSharingConfiguration(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        let fromAppGroup: String? = call.getParameter(Args.fromAppGroup)
        let toAppGroup: String? = call.getParameter(Args.toAppGroup)

        // Resolve the source and destination UserDefaults. A nil app group means the standard,
        // non-shared UserDefaults; a non-nil app group means the shared suite for that group.
        guard let from = userDefaults(forAppGroup: fromAppGroup),
              let to = userDefaults(forAppGroup: toAppGroup) else {
            // One of the suites could not be opened - data sharing is probably not configured properly.
            throw PluginException(.wrongParameter, message: "Failed to open UserDefaults for the provided app group. Data sharing is probably not configured properly.")
        }

        // If the destination is already initialized, the migration is not required.
        if to.bool(forKey: PowerAuthKeychain_Initialized) {
            result(nil)
            return
        }

        // Move the flag from the source suite to the destination suite.
        if from.bool(forKey: PowerAuthKeychain_Initialized) {
            from.removeObject(forKey: PowerAuthKeychain_Initialized)
            from.synchronize()
            to.set(true, forKey: PowerAuthKeychain_Initialized)
            to.synchronize()
        }

        result(nil)
    }

    /// Returns the `UserDefaults` for the given app group, or the standard `UserDefaults` when
    /// [appGroup] is `nil`. Returns `nil` when a shared suite for the given app group cannot be opened.
    private func userDefaults(forAppGroup appGroup: String?) -> UserDefaults? {
        guard let appGroup else {
            return .standard
        }
        return UserDefaults(suiteName: appGroup)
    }
}
